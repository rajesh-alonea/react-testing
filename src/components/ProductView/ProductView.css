.product {
  margin-bottom: var(--sp-2);
  padding: var(--sp-2);
  border-radius: var(--px-4);
  cursor: pointer;
  user-select: none;
}

.product__photo {
  object-fit: cover;
  height: 100%;
  width: 10rem;
}
