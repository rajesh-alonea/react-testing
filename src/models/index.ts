export * from './Address';
export * from './Env';
export * from './CheckoutInfo';
export * from './Cart';
export * from './Order';
export * from './OrderItem';
export * from './Product';
