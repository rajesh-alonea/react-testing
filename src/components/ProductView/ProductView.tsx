import React from 'react';
import { NumberUtils } from '@react-force/number-utils';
import { Product } from '../../models';
import { HorizontalContainer } from '../Containers';
import './ProductView.css';

export interface ProductViewProps {
  product: Product;
  onClick: (productId: string) => void;
}

export const ProductView = ({ product, onClick }: ProductViewProps) => {
  const { id, name, description, price, photo } = product;

  return (
    <HorizontalContainer
      testId="product"
      className="product paper border-paper items-center"
      onClick={() => onClick(id)}
    >
      <img className="product__photo" src={photo} alt={name} />
      <div className="ml-4">
        <h3>{name}</h3>
        <p className="mt-0">{description}</p>
        <p className="mb-0">${NumberUtils.formatAsMoney(price)}</p>
      </div>
    </HorizontalContainer>
  );
};
