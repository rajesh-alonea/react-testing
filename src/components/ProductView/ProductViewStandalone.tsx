import React from 'react';
import { NumberUtils } from '@react-force/number-utils';
import { Product } from '../../models';
import { CartService } from '../../services';
import { HorizontalContainer } from '../Containers';
import './ProductView.css';

export interface ProductViewStandaloneProps {
  product: Product;
}

/** Standalone version of ProductView which handles the onClick internally */
export const ProductViewStandalone = ({
  product,
}: ProductViewStandaloneProps) => {
  const { id, name, description, price, photo } = product;

  const handleClick = async () => {
    await CartService.addProduct(id);
  };

  return (
    <HorizontalContainer
      testId="product"
      className="product paper border-paper items-center"
      onClick={handleClick}
    >
      <img className="product__photo" src={photo} alt={name} />
      <div className="ml-4">
        <h3>{name}</h3>
        <p className="mt-0">{description}</p>
        <p className="mb-0">${NumberUtils.formatAsMoney(price)}</p>
      </div>
    </HorizontalContainer>
  );
};
