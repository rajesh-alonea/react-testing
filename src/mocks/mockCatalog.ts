import { Catalog } from '../models';

export const mockCatalog: Catalog = {
  'apple-imac': {
    id: 'apple-imac',
    name: 'iMac',
    description:
      'The 27‑inch iMac is packed with powerful tools and apps that let you take any idea to the next level. Its superfast processors and graphics, massive memory, and all-flash storage can tackle any workload with ease. And with its advanced audio and video capabilities and stunning 5K Retina display, everything you do is larger than life.',
    manufacturer: 'Apple',
    price: 1299,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/apple-imac.jpg?alt=media&token=69d03695-d57d-4b1a-952c-909a21af9099',
  },
  'apple-macbook-pro': {
    id: 'apple-macbook-pro',
    name: 'MacBook Pro',
    description:
      'Designed for those who defy limits and change the world, the 16-inch MacBook Pro is by far the most powerful notebook Apple have ever made. With an immersive Retina display, superfast processors, advanced graphics, the largest battery capacity ever in a MacBook Pro, Magic Keyboard, and massive storage.',
    manufacturer: 'Apple',
    price: 699,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/apple-macbook-pro.jpg?alt=media&token=d2b1cf37-cfe0-4f20-adeb-eaeb472cedaa',
  },
  'apple-ipad': {
    id: 'apple-ipad',
    name: 'iPad',
    description:
      'Powerful. Easy to use. Versatile. The new iPad is designed for all the things you love to do. Work, play, create, learn, stay connected, and more. All at an incredible value.',
    manufacturer: 'Apple',
    price: 249,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/apple-ipad.jpg?alt=media&token=17ba3b67-5e54-4bf0-8dfb-afb2538355d2',
  },
  'belkin-netcam': {
    id: 'belkin-netcam',
    name: 'Belkin NetCam HD+ Wi-Fi Camera',
    description:
      "The Belkin NetCam HD+ delivers 720p HD live streaming video to your smartphone or tablet, letting you watch the action in your home from wherever you go. Equipped with a wide-angle glass lens, night vision, and secure Wi-Fi connectivity, the NetCam HD+ ensures you'll always be a part of what's happening at home.",
    manufacturer: 'Belkin',
    price: 99,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/belkin-netcam-hd%2B-wi-fi-camera.jpg?alt=media&token=26babba0-3765-4402-8336-facf19c07ca7',
  },
  'fitbit-charge-hr': {
    id: 'fitbit-charge-hr',
    name: 'Fitbit Charge HR',
    description:
      'Make every beat count with Charge HR—a high-performance wristband with automatic, continuous heart rate and activity tracking right on your wrist.',
    manufacturer: 'Fitbit',
    price: 136,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/fitbit-charge-hr.jpg?alt=media&token=ba217f84-6b73-4a45-8c41-eaf75242668f',
  },
  'ring-video-doorbell-pro': {
    id: 'ring-video-doorbell-pro',
    name: 'Ring Video Doorbell Pro',
    description:
      "Our wired doorbell camera with added security and style. Video Doorbell Pro brings you next-level features like enhanced dual-band wifi, color night vision, and built-in Alexa Greetings that answer the door for you when you're busy.",
    manufacturer: 'Ring',
    price: 249,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/ring-video-doorbell-pro.jpg?alt=media&token=91de25d4-813f-4341-9fec-cd8564ae4335',
  },
  'august-smart-lock-pro': {
    id: 'august-smart-lock-pro',
    name: 'August Smart Lock Pro',
    description:
      'The ultimate smart door upgrade for your home. August Smart Lock Pro takes any worry out of getting into your home. Use our top-rated app to control your door to unlock/lock, grant guest access, see who came and left, and let anyone in from anywhere.',
    manufacturer: 'August',
    price: 249,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/august-smart-lock-pro.jpg?alt=media&token=026b30c5-6b4f-4432-9e8e-5c0830274bee',
  },
  'nest-learning-thermostat': {
    id: 'nest-learning-thermostat',
    name: 'Nest Learning Thermostat',
    description:
      'The Nest Learning Thermostat automatically adapts as your life and the seasons change. Just use it for a week and it programs itself. With the Nest app, your thermostat lives on your wall and in your pocket. Once you connect your thermostat to Wi-Fi, you can control it from anywhere.',
    manufacturer: 'Nest',
    price: 249,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/nest-learning-thermostat.jpg?alt=media&token=afea31e6-611e-4ec4-a2fe-6cb9928ab997',
  },
  'nest-protect': {
    id: 'nest-protect',
    name: 'Nest Protect Smoke + Carbon Monoxide Alarm',
    description:
      'There’s never been a smoke and CO alarm quite like this. Nest Protect looks for fast-burning fires, smoldering fires, and carbon monoxide. It speaks up when there’s a problem and can send a message to your phone if you’re not at home. And it helps keep you safe and sound for up to a decade.',
    manufacturer: 'Nest',
    price: 249,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/nest-protect.jpg?alt=media&token=c36ef284-bd4b-4cfb-90e2-1fd5e59417d3',
  },
  'google-chromecast-audio': {
    id: 'google-chromecast-audio',
    name: 'Google Chromecast Audio',
    description:
      "Stream your favorite music with Google Chromecast Audio. Simply plug into your speaker's auxiliary input and connect to your Wi-Fi network. Then just tap the Cast button from your favorite apps on your phone, tablet or computer to cast songs, radio stations and podcasts to your speaker.",
    manufacturer: 'Google',
    price: 35,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/google-chromecast-audio.jpg?alt=media&token=83800a2b-575a-4927-8c9c-2e16f25e87a9',
  },
  'nest-secure': {
    id: 'nest-secure',
    name: 'Nest Secure Alarm System Starter Pack',
    description:
      'Tough on bad guys. Easy on you. The Nest secure alarm system is easy to live with every day. Just tap Nest tag on Nest guard to arm and disarm the alarm – no pass code needed. Nest detect sensors look out for doors, windows and rooms. And with the Nest app, you’ll always know what’s happening at home.',
    manufacturer: 'Nest',
    price: 499,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/nest-secure.jpg?alt=media&token=4dc8e77b-e94e-416c-ad6c-3182397264a6',
  },
  'sonos-play1': {
    id: 'sonos-play1',
    name: 'Sonos PLAY:1 Smart Wireless Speaker',
    description:
      'Small yet powerful speaker for streaming music and more. Get rich, room-filling sound with Play:1, and control it with the Sonos app.',
    manufacturer: 'Sonos',
    price: 149,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/sonos-play1.jpg?alt=media&token=9e04f562-5d11-4b0d-a5d2-7c7c40137fc6',
  },
  'dyson-cordless-vacuum': {
    id: 'dyson-cordless-vacuum',
    name: 'Dyson Cordless Vacuum',
    description:
      'For cord-free cleaning with twice the suction of any other cordless stick vacuum cleaner.',
    manufacturer: 'Dyson',
    price: 175,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/dyson-cordless-vacuum.jpg?alt=media&token=0ca780cf-475f-4d82-8486-5ae36696421f',
  },
  'gopro-hero': {
    id: 'gopro-hero',
    name: 'GoPro HERO Session Action Camera',
    description:
      'HERO Session packs the power of GoPro into a convenient, grab-and-go, everyday camera. Perfect for the first-time GoPro user, or as a second camera, HERO Session is simple and easy to use. A single button powers on the camera and starts recording automatically, then when you stop recording, the camera powers itself off.',
    manufacturer: 'GoPro',
    price: 149,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/gopro-hero.jpg?alt=media&token=22e54972-b834-4319-be7c-f1d046e7487f',
  },
  'google-home-mini': {
    id: 'google-home-mini',
    name: 'Google Home Mini',
    description:
      'Google Home Mini is a smart speaker powered by the Google Assistant, so you can ask it questions and tell it to do things. Just start with "Ok Google" to enjoy your entertainment, get answers from Google, and control your smart home.',
    manufacturer: 'Google',
    price: 29,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/google-home-mini.jpg?alt=media&token=af98464b-f994-4473-b96b-7549c375b6c1',
  },
  'canon-eos-rebel': {
    id: 'canon-eos-rebel',
    name: 'Canon EOS Rebel SL1',
    description:
      "As Canon's smallest and lightest digital SLR camera, the EOS Rebel SL1 is small in size but enormous in performance. With a newly-designed Canon 18.0 Megapixel CMOS (APS-C) sensor and speedy Canon DIGIC 5 Image Processor, it delivers images of extraordinary quality.",
    manufacturer: 'Canon',
    price: 403,
    photo:
      'https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/canon-eos-rebel.jpg?alt=media&token=aaab2139-a9ba-4ac1-a9bc-a4902a84926c',
  },
};
