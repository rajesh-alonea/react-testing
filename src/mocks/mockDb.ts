import { Storage } from '../utils';
import { Cart } from '../models';

const CART_KEY = 'mockDbCart';

// The mock database keeps all data in memory. However it is backed by
// localStorage. Anytime a value is written to the in-memory database,
// it is also persisted to localStorage.

// -------------------- Initialize in-memory database --------------------
// cart
let cart: Cart = Storage.get(CART_KEY, { items: [] });
// -----------------------------------------------------------------------

function getCart(): Cart {
  return cart;
}

function setCart(newCart: Cart) {
  cart = newCart;
  Storage.set(CART_KEY, cart);
}

function clearCart() {
  cart = { items: [] };
  Storage.set(CART_KEY, cart);
}

export const mockDb = {
  getCart,
  setCart,
  clearCart,
};
