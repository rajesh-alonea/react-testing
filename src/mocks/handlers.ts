import { rest } from 'msw';
import { v4 as uuidv4 } from 'uuid';
import { CartUtils, CheckoutInfo, Order } from '../models';
import { MOCK_API_URL } from './constants';
import { mockCatalog } from './mockCatalog';
import { mockDb } from './mockDb';
import mockOrders from './mockOrders.json';

interface AddProductInput {
  productId: string;
}

export const handlers = [
  /** get catalog */
  rest.get(`${MOCK_API_URL}/catalog`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockCatalog));
  }),

  /** get cart */
  rest.get(`${MOCK_API_URL}/cart`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockDb.getCart()));
  }),

  /** get orders */
  rest.get(`${MOCK_API_URL}/orders`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockOrders));
  }),

  /** add a product to the cart */
  rest.post(`${MOCK_API_URL}/cart/items`, (req, res, ctx) => {
    const { productId } = req.body as AddProductInput;
    const cart = mockDb.getCart();
    const newCart = CartUtils.addProduct(cart, mockCatalog[productId]);
    mockDb.setCart(newCart);
    return res(ctx.status(200), ctx.json(newCart));
  }),

  /** delete an item from the cart */
  rest.delete(`${MOCK_API_URL}/cart/items/:productId`, (req, res, ctx) => {
    const { productId } = req.params;
    const cart = mockDb.getCart();
    const newCart = CartUtils.deleteItem(cart, productId);
    mockDb.setCart(newCart);
    return res(ctx.status(200), ctx.json(newCart));
  }),

  /** set the quantity of an item in the cart **/
  rest.patch(`${MOCK_API_URL}/cart/items/:productId`, (req, res, ctx) => {
    const { productId } = req.params;
    const { quantity } = req.body as any;
    const cart = mockDb.getCart();
    const newCart = CartUtils.setItemQuantity(cart, productId, quantity);
    mockDb.setCart(newCart);
    return res(ctx.status(200), ctx.json(newCart));
  }),

  /** create an order */
  rest.post(`${MOCK_API_URL}/orders`, (req, res, ctx) => {
    const checkoutInfo = req.body as CheckoutInfo;

    // Move cart items into the order
    const order: Order = {
      id: uuidv4(),
      createdAt: new Date().toISOString(),
      items: mockDb.getCart().items,
      shippingAddress: checkoutInfo.shippingAddress,
    };
    // @ts-ignore
    mockOrders.push(order);

    // clear the cart
    mockDb.clearCart();

    return res(ctx.status(200), ctx.json(order));
  }),
];
