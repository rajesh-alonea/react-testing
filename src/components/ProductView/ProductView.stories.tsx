import React from 'react';
import { action } from '@storybook/addon-actions';
import { Meta, Story } from '@storybook/react';
import { mockCatalog } from '../../mocks/mockCatalog';
import { Product } from '../../models';
import { ProductView } from './ProductView';

const product: Product = mockCatalog['apple-imac'];

export default {
  title: 'Components/ProductView',
  component: ProductView,
} as Meta;

const Template: Story = (args) => (
  <ProductView product={args.product} onClick={action('Product clicked')} />
);

export const ProductViewStory = Template.bind({});
ProductViewStory.storyName = 'ProductView';
ProductViewStory.args = { product };
