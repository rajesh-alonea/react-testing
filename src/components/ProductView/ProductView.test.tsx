import React from 'react';
import { render, screen, userEvent } from '../../test/test-utils';
import { mockCatalog } from '../../mocks/mockCatalog';
import { Product } from '../../models';
import { ProductView } from './ProductView';

const product: Product = mockCatalog['apple-imac'];
const handleClick = jest.fn();

describe('<ProductView />', () => {
  test('renders correctly', async () => {
    const { asFragment } = render(
      <ProductView product={product} onClick={handleClick} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('when clicked, calls onClick with productId', async () => {
    render(<ProductView product={product} onClick={handleClick} />);

    // click on the ProductView
    userEvent.click(screen.getByTestId('product'));

    // expect mock handler to be called
    expect(handleClick).toBeCalledTimes(1);
    expect(handleClick).toBeCalledWith(product.id);
  });
});
