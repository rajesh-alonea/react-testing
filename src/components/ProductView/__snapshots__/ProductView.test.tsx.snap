// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductView /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="flex-1 flex flex-row product paper border-paper items-center"
    data-testid="product"
  >
    <img
      alt="iMac"
      class="product__photo"
      src="https://firebasestorage.googleapis.com/v0/b/mobx-shop.appspot.com/o/apple-imac.jpg?alt=media&token=69d03695-d57d-4b1a-952c-909a21af9099"
    />
    <div
      class="ml-4"
    >
      <h3>
        iMac
      </h3>
      <p
        class="mt-0"
      >
        The 27â€‘inch iMac is packed with powerful tools and apps that let you take any idea to the next level. Its superfast processors and graphics, massive memory, and all-flash storage can tackle any workload with ease. And with its advanced audio and video capabilities and stunning 5K Retina display, everything you do is larger than life.
      </p>
      <p
        class="mb-0"
      >
        $1,299.00
      </p>
    </div>
  </div>
</DocumentFragment>
`;
